syncopy/preproc/firws.py

```diff
@@ -0,0 +1,237 @@
+    lp_kernel = windowed_sinc(window, order, f_lp)
+    kernel = windowed_sinc(window, order, f_hp)
+    hp_kernel = invert_sinc(kernel)
+    kernel = lp_kernel + hp_kernel
+
+    # subtract dc component from filter addition
+    # for band-pass filters in the time-domain
+    if filter_type == 'bp':
+        kernel[len(kernel) // 2] -= 1
+
+    return kernel
+
+
+def windowed_sinc(window, order, f_c):
+
+    """
+    Construct the symmetric windowed sinc filter
+    with a cut-off frequency `f_c`
+
+    Parameters
+    ----------
+    window : str
+        One of `scipy.signal.windows`
+    order : int
+       The order of the filter, has to be strictly even
+    f_c : float
+       Cut-off frequency in sampling units,
+       maximum is Nyquist `f_c=0.5`
+
+    Returns
+    -------
+    kernel : :class:`numpy.ndarray`
+        The windowed filter with length `order + 1`
+
+    """
+
+    # angular cut-off frequency
+    omega_c = 2 * np.pi * f_c
+
+    win_func = getattr(sci_win, window)
+    win = win_func(order + 1)
+
+    # one-sided support
+    m_half = np.arange(1, order / 2 + 1)
+    kernel = np.sin(omega_c * m_half) / m_half
+    kernel = np.hstack([kernel[::-1], omega_c, kernel]) * win
+    # normalize to unity gain
+    kernel = kernel / kernel.sum()
+
+    return kernel
+
+
+def invert_sinc(kernel):
+
+    """
+    In frequency space the high-pass filter
+    is just 1 - low-pass. Hence, formally we can
+    calculate the high-pass version
+    in the time domain via the correspinding inverse Fourier:
+
+    Fourier^-1 {1 - Fourier(kernel)} = Fourier^-1(1) - kernel
+
+    This gives a delta-peak at the 0-lag midpoint
+    minus the original low-pass impulse response.
+    In the DSP world the delta distribution corresponds to a mere `1`.
+    """
+    kernel = -kernel
+    # kernel size is always odd
+    kernel[len(kernel) // 2] += 1
+    return kernel
+
+
+def minphaserceps(fkernel):
+
+    """
+    Tranform FIR filter to minmum phase (causal) filter
+
+    The original Matlab function was written for FieldTrip in 2013 by
+    Andreas Widmann, University of Leipzig, widmann@uni-leipzig.de
+
+    Notes
+    -----
+    .. [1] Smith III, O. J. (2007). Introduction to Digital Filters with Audio
+       Applications. W3K Publishing. Retrieved Nov 11 2013, from
+       https://ccrma.stanford.edu/~jos/fp/Matlab_listing_mps_m.html
+    .. [2] Vetter, K. (2013, Nov 11). Long FIR filters with low latency.
+       Retrieved Nov 11 2013, from
+       http://www.katjaas.nl/minimumphase/minimumphase.html
+    """
+
+    nSamples = len(fkernel)
+    upsamplingFactor = 1e3  # Impulse response upsampling/zero padding to reduce time-aliasing
+    nFFT = int(2**np.ceil(np.log2(nSamples * upsamplingFactor)))  # Power of 2
+    clipThresh = 1e-8  # -160 dB
+
+    # Spectrum
+    specC = np.abs(np.fft.fft(fkernel, nFFT))
+    specC[specC < clipThresh] = clipThresh  # Clip spectrum to reduce time-aliasing
+
+    # Real cepstrum
```

Comment:
`cepstrum`?

Reply:
Yap, `cepstrum` is what's  the log-transformed amplitude spectrum is called :man_shrugging:

Reply:
Maybe  typo, and it's actually `spectrum`?

Reply:
Ok, I now learned about the [Cepstrum](https://en.wikipedia.org/wiki/Cepstrum)